symbolic-base/src/ZkFold/Base/Protocol/Protostar/AccumulatorScheme.hs

```diff
@@ -22,94 +23,75 @@ import           ZkFold.Base.Protocol.Protostar.CommitOpen   (CommitOpen (..), C
--- linearCombination a1 a2 -> a1 * X + a2
---
-class LinearCombination a b where
-    linearCombination :: a -> a -> b
-
--- | Same as above, but with a coefficient known at runtime
--- linearCombination coeff b1 b2 -> b1 * coeff + b2
---
-class LinearCombinationWith a b where
-    linearCombinationWith :: a -> b -> b -> b
+           -> f                           -- Commitment key ck
+           -> Accumulator pi f c m        -- final accumulator
+           -> ([c], c)                    -- returns zeros if the final accumulator is valid
 
-instance (Scale f a, AdditiveSemigroup a) => LinearCombinationWith f [a] where
-    linearCombinationWith f = P.zipWith (\a b -> scale f a + b)
+type DataRepresentableAsList f x = (IsList x, Item x ~ f)
 
 instance
-    ( Symbolic ctx
-    , Eq (Bool ctx) c
-    , Eq (Bool ctx) i
-    , Eq (Bool ctx) f
-    , Eq (Bool ctx) [f]
-    , Eq (Bool ctx) [c]
+    ( AdditiveGroup pi
     , AdditiveGroup c
     , AdditiveSemigroup m
     , Ring f
+    , Scale f pi
     , Scale f c
     , Scale f m
-    , MapInput f a ~ i
-    , deg ~ Degree (CommitOpen m c a) + 1
-    , KnownNat deg
-    , LinearCombination (MapMessage f a) (MapMessage (PU.PolyVec f deg) a)
-    , LinearCombination (MapInput f a) (MapInput (PU.PolyVec f deg) a)
-    , LinearCombinationWith f (MapInput f a)
-    , MapMessage f a ~ m
+    , RandomOracle pi f         -- Random oracle for compressing public input
+    , RandomOracle c f          -- Random oracle ρ_NARK
+    , HomomorphicCommit f [f] c
+    , HomomorphicCommit f m c
+    , IsList pi
+    , Item pi ~ f
+    , IsList m
+    , Item m ~ f
     , AlgebraicMap f (CommitOpen m c a)
+    , MapInput f a ~ pi
+    , MapMessage f a ~ m
+    , Degree (FiatShamir f (CommitOpen m c a)) + 1 ~ deg
+    , KnownNat deg
     , AlgebraicMap (PU.PolyVec f deg) a
-    , RandomOracle c f                                    -- Random oracle ρ_NARK
-    , RandomOracle i f                                    -- Random oracle for compressing public input
-    , HomomorphicCommit f [m] c
-    , HomomorphicCommit f [f] c
-    ) => AccumulatorScheme i f c m ctx (FiatShamir f (CommitOpen m c a)) where
-  prover (FiatShamir (CommitOpen _ sps) _) ck acc (InstanceProofPair pubi (NARKProof pi_x pi_w)) =
-        (Accumulator (AccumulatorInstance pi'' ci'' ri'' eCapital' mu') mi'', eCapital_j)
+    , MapInput (PU.PolyVec f deg) a ~ [PU.PolyVec f deg]
+    , MapMessage (PU.PolyVec f deg) a ~ PU.PolyVec f deg
+    ) => AccumulatorScheme pi f c m (FiatShamir f (CommitOpen m c a)) where
+  prover (FiatShamir (CommitOpen _ sps)) ck acc (InstanceProofPair pubi (NARKProof pi_x pi_w)) =
+        (Accumulator (AccumulatorInstance pi'' ci'' ri'' eCapital' mu') m_i'', pf)
       where
           -- Fig. 3, step 1
           r_i :: [f]
-          r_i = P.tail $ P.scanl (P.curry oracle) (oracle pubi) (zero : pi_x)
+          r_i = P.scanl (P.curry oracle) (oracle pubi) pi_x
```

Comment:
Why this change? I understand that these are random challenges, and it's probably unimportant how they are calculated, but the old implementation corresponds to the paper

Reply:
Hmm, I don't see how the previous version corresponds to the paper. On the first iteration, you were calculating `oracle (oracle pubi) zero` and using it in the calculation for $r_i$. In the paper, on the first iteration, you do `oracle (oracle pubi) c_1` where `c_1` is the commitment of the first message. But I agree that I should take the `tail` as before.

Reply:
Well, in my opinion, r0 := ρNARK(pi) in the paper corresponds to just `oracle pubi`. Either way, you need two oracles while in the paper they are denoted the same.